src/python/pants/core/goals/deploy.py

```diff
@@ -115,51 +128,294 @@ async def publish_process_for_target(
+    def closure(self) -> Iterator[_DeployStepResult]:
+        for substep in self.substeps:
+            yield from substep.closure
+
+        yield self
+
+
+class _DeployStepResults(Collection[_DeployStepResult]):
+    @memoized_property
+    def exit_code(self) -> int:
+        if len(self) == 0:
+            return 0
+
+        return max(result.exit_code for result in self)
+
+    @property
+    def closure(self) -> Iterator[_DeployStepResult]:
+        for result in self:
+            yield from result.closure
+
+
+@rule(desc="Publish package dependencies", level=LogLevel.DEBUG)
+async def _run_publish_process(publish: PublishPackages) -> _DeployStepResult:
+    if publish.process:
+        # TODO cheating here, but only by invoking the underly process I can run this in a `@rule`
+        result = await Get(FallibleProcessResult, Process, publish.process.process)
+        return _DeployStepResult.from_publish_process_result(result, publish)
+
+    return _DeployStepResult.skipped(
+        action=_DeployStepAction.PUBLISH_PACKAGE,
+        names=publish.names,
+        description=publish.description,
+    )
+
+
+@rule(desc="Run deployment step", level=LogLevel.DEBUG)
+async def _run_deploy_process(deploy: DeployProcess) -> _DeployStepResult:
+    publish_results: Iterable[_DeployStepResult] = []
+    if deploy.publish_dependencies:
+        publish_processes = await _publish_processes_for_targets(deploy.publish_dependencies)
+        publish_results = _DeployStepResults(
+            await MultiGet(
+                Get(_DeployStepResult, PublishPackages, process) for process in publish_processes
+            )
+        )
+
+        if publish_results.exit_code != 0:
+            return _DeployStepResult.failed_dependencies(deploy, publish_results)
+
+    if deploy.process:
+        # TODO cheating here, but only by invoking the underly process I can run this in a `@rule`
+        result = await Get(FallibleProcessResult, Process, deploy.process.process)
+        return _DeployStepResult.from_deploy_process_result(
+            result,
+            deploy,
+            substeps=publish_results,
+        )
+
+    return _DeployStepResult.skipped(
+        action=_DeployStepAction.DEPLOYMENT,
+        names=(deploy.name,),
+        description=deploy.description,
+        substeps=publish_results,
+    )
+
+
+@rule(desc="Deploy target dependencies", level=LogLevel.DEBUG)
+async def _deploy_target_dependencies(
+    request: _DeployTargetDependenciesRequest,
+) -> _DeployTargetRequests:
+    return _DeployTargetRequests(
+        [
+            _DeployTargetRequest(dep, dependent=request.target.representative.address)
+            for dep in request.target.dependencies
+        ]
+    )
+
+
+@rule
+async def _deploy_targets(requests: _DeployTargetRequests) -> _DeployStepResults:
+    result = await MultiGet(
+        Get(_DeployStepResult, _DeployTargetRequest, request) for request in requests
+    )
+    return _DeployStepResults(result)
```

Comment:
It seems like these two rules could be merged (and the `_DeployTargetRequests` type eliminated).

Reply:
I was experiencing an infinite loop when invoking the `MultiGet(_DeployStepResult, ...)` from inside the other rule, splitting them into two rules cleared that out although I may have been doing something wrong.